tracer/src/Datadog.Trace/Ci/Agent/CIWriterHttpSender.cs

```diff
@@ -0,0 +1,166 @@
+// <copyright file="CIWriterHttpSender.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+using Datadog.Trace.Agent;
+using Datadog.Trace.Agent.Transports;
+using Datadog.Trace.Ci.Agent.Payloads;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Ci.Agent
+{
+    internal sealed class CIWriterHttpSender : ICIAgentlessWriterSender
+    {
+        private const string ApiKeyHeader = "dd-api-key";
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<CIWriterHttpSender>();
+
+        private readonly IApiRequestFactory _apiRequestFactory;
+
+        public CIWriterHttpSender(IApiRequestFactory apiRequestFactory)
+        {
+            _apiRequestFactory = apiRequestFactory;
+            Log.Information("CIWriterHttpSender Initialized.");
+        }
+
+        public async Task SendPayloadAsync(EventsPayload payload)
+        {
+            var numberOfTraces = payload.Count;
+            var tracesEndpoint = payload.Url;
+
+            // retry up to 5 times with exponential back-off
+            const int retryLimit = 5;
+            var retryCount = 1;
+            var sleepDuration = 100; // in milliseconds
+
+            var payloadMimeType = MimeTypes.MsgPack;
+            var payloadBytes = payload.ToArray();
+
+            Log.Information($"Sending ({numberOfTraces} events) {payloadBytes.Length.ToString("N0")} bytes...");
+
+            while (true)
+            {
+                IApiRequest request;
+
+                try
+                {
+                    request = _apiRequestFactory.Create(tracesEndpoint);
+                    request.AddHeader(ApiKeyHeader, CIVisibility.Settings.ApiKey);
+                }
+                catch (Exception ex)
+                {
+                    Log.Error(ex, "An error occurred while generating http request to send events to {AgentEndpoint}", _apiRequestFactory.Info(tracesEndpoint));
+                    return;
+                }
+
+                bool success = false;
+                Exception exception = null;
+                bool isFinalTry = retryCount >= retryLimit;
+
+                try
+                {
+                    success = await SendPayloadAsync(new ArraySegment<byte>(payloadBytes), payloadMimeType, request, isFinalTry).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    exception = ex;
+#if DEBUG
```

Comment:
[non-blocking]
I know we do this in the tracer, but I'm not a big fan of using compile-time properties for this. Can we check if debug mode is enabled instead?

Reply:
Sure, will do.

Reply:
done in [90dbbc1](https://github.com/DataDog/dd-trace-dotnet/pull/2402/commits/90dbbc15313b0a8251edf04da63ae09a4a4ded94)